Deploy/InitTerraform.ps1

```diff
@@ -0,0 +1,119 @@
+<#
+.SYNOPSIS
+Init the remote state store used by Terraform (Blob Storage) and the execution folders
+
+.DESCRIPTION
+Init the remote state store used by Terraform (Blob Storage) and the execution folders
+
+This script will do following steps:
+
+1. Ensure Resource group for Terraform
+2. Ensure Storage Account and Container for Terraform Remote State
+3. Initalizes all Terraform folders
+
+After the script is successfully executed the Terraform can use the storage as remote state store.
+
+.EXAMPLE
+.\InitTerraform.ps1 -STORAGE_ACCOUNT_NAME myterraformstate -RESOURCE_GROUP_NAME rg-myterraformstate
+
+.INPUTS
+None. You cannot pipe objects.
+
+.OUTPUTS
+System.Boolean. Returns $True if executed successfully
+
+#>
+param(
+    # Storage Account Name for Terraform remote state
+    [Parameter(Mandatory=$true, HelpMessage="Storage Account Name for Terraform remote state")]
+    [string] $STORAGE_ACCOUNT_NAME,
+
+    # Resource Group Name where Storage Account is placed for Terraform remote state
+    [Parameter(Mandatory=$true, HelpMessage="Resource Group Name where Storage Account is placed for Terraform remote state")]
+    [string] $RESOURCE_GROUP_NAME,
+
+    # Region used for Resource Group and Storage Account
+    [Parameter(HelpMessage="Region used for global services")]
+    [string] $LOCATION = "japaneast",
+
+    # Terraform folders from within Deploy folder
+    [Parameter(HelpMessage="Terraform folders from within Deploy folder")]
+    [string[]] $TERRAFORM_FOLDERS =  @("IaC", "SSLActivation", "SSLIssuing"),
+
+    # Maximum wait time for RBAC rights to be propagated. Default 3 min
+    [Parameter(HelpMessage="Maximum wait time for RBAC rights to be propagated. Default 3 min")]
+    [int] $MAX_WAIT_TIME_MIN = 3
+)
+# Import Helper functions
+. "$($MyInvocation.MyCommand.Path -replace($MyInvocation.MyCommand.Name))\HelperFunctions.ps1"
+# Helper var
+$success = $True
+$loopcount = 0
+$waitretrysec = 10
+$loopmax = (60 * $MAX_WAIT_TIME_MIN ) / $waitretrysec
+$container_name = "tfstate"
+
+# Tell who you are (See HelperFunction.ps1)
+Write-WhoIAm
+
+# 1. Ensure Resource group for Terraform
+Write-Host "## 1. Ensure Resource group for Terraform"
+az group create --name $RESOURCE_GROUP_NAME --location $LOCATION > $null
+$success = $success -and $?
+
+# 2. Ensure Storage Account and Container for Terraform Remote State
+Write-Host "## 2. Ensure Storage Account and Container for Terraform Remote State"
+$account = az storage account create --resource-group $RESOURCE_GROUP_NAME --name $STORAGE_ACCOUNT_NAME --kind StorageV2 --sku Standard_LRS --https-only true | ConvertFrom-Json
+$success = $success -and $?
+
+# Get info around current connection
+$currentConnection = az account show | ConvertFrom-Json
+$success = $success -and $?
+
+# Save Principal Type and Id
+$principalType = $currentConnection.user.type
+$principalId = ""
+
+if($principalType -eq "user")
```

Comment:
Does the behavior differ depending on what the 'service principal type' is? Or is there only difference command which should be executed according to 'service principal type'?

Reply:
You can log into AzureCLI as a user or with a service principal. We need e.g. the service principal login for our test pipeline.
In order to give access rights on the Storage Account for the Terraform State Store we need the AAD ObjectId. However the way how we can retrieve the 'ObjectId' of a **user** or a **service principal** is different. So I try to distinct what login type is used and retrieve the objectId accordingly.